tincr/io/library/genCellLibrary.tcl

```diff
@@ -300,275 +727,368 @@ proc process_macrocell {c s fo} {
+        puts $xml_out "        </is_lut>"
+    }
+    
+    # Tag VCC and GND cells
+    if { $lib_cell_name == "VCC" } {
+        puts $xml_out "        <vcc_source/>"
+    }
+    if { $lib_cell_name == "GND" } {
+        puts $xml_out "        <gnd_source/>"
+    }
+
+    # Tag the primitive level (currently unused)
+    puts $xml_out "      <level>[get_property PRIMITIVE_LEVEL $cell_instance]</level>"
+}
+
+## Writes the cell pin information to the specified XML file. 
+#   The name, direction, and type (see {@link get_pin_type}) is
+#   printed for each pin.
+#
+# @param cell_instance Library cell instance
+# @param xml_out File handle to the output XML file
+proc write_pin_xml { cell_instance xml_out } {
+    
+    puts $xml_out "      <pins>"
+
+    #print the cell pin information
+    foreach cell_pin [get_pins -of $cell_instance] {
+        puts $xml_out "        <pin>"
+        puts $xml_out "          <name>[get_property REF_PIN_NAME $cell_pin]</name>"
+
+        set dir [get_property DIRECTION $cell_pin]
+        if { $dir == "IN" } then {
+            puts $xml_out "          <direction>input</direction>"
+        } elseif { $dir == "OUT" }  {
+            puts $xml_out "          <direction>output</direction>"
+        } else {
+            puts $xml_out "          <direction>inout</direction>"
         }
-        
+        puts $xml_out "          <type>[get_pin_type $cell_pin]</type>"
+
+        puts $xml_out "        </pin>"
     }
-    puts $fo "      </bels>"
-    puts $fo "    </cell>"
+    puts $xml_out "      </pins>"
 }
 
-# Go through and print out all the properties associated with a library cell
-proc print_libcell_properties {fo lc } {
-    set pl [list_property $lc]
-    set lst [list]
-    foreach p $pl  {
-	if { [string first "CONFIG." $p] == 0 } {
-	    set tmp [split $p "."]
-	    if { [llength $tmp] == 2 } {
-		lappend lst $p
-	    }
-	}
-
-	# Add the non-CONFIG properties we care about
-	if { $p == "PRIMITIVE_GROUP" } {
-	    lappend lst $p
-	}
-    }
-    
-    if { [llength $lst] > 0 } {
-	puts $fo "      <libcellproperties>"
-	foreach h $lst {
-	    puts $fo "        <libcellproperty>"
-	    puts $fo "          <name>$h</name>"
-
-	    # For non-CONFIG properties
-	    if { [string first "CONFIG" $h]  == -1 } {
-		puts $fo "          <value>[get_property $h $lc]</value>"
-		puts $fo "          <readonly/>"
-	    }
-	    
-	    foreach p $pl {
-		if { [llength [split $p "."]] == 3 && [string first $h $p] == 0 } {
-		    set tag [lindex [split $p "."] 2]
-		    set tag [string tolower $tag]
-		    puts $fo "          <$tag>[get_property $p $lc]</$tag>"
-		}
-	    }
-	    puts $fo "        </libcellproperty>"
-	}
-	puts $fo "      </libcellproperties>"
+## Gets the type of the specified pin according to Vivado. DATA is the
+#   default pin type where no other pin type is specified.
+#   TODO: on each new release of Vivado, verify this function is still correct.
```

Comment:
Unfortunately, the correctness of this function has to be manually verified. This is because for a cell pin, there are several pin types (as shown in the function implementation) which are based on the naming "IS_*" (for example IS_CLOCK). However, there are also other properties such as IS_CONNECTED, IS_LEAF, and IS_ORIG_PIN that are not pin types but match the pattern. There is no way to reliable filter these out in a test. I added a not in the comments saying that it has to be manually verified.